Fix python::pip installing $editable VCS packages every Puppet run

wolttam opened the pull request:
#### Pull Request (PR) description
This is due to VCS packages appearing in `pip freeze -all` as:
-e git://example.com/package.git@<ref>#egg=<egg>

Updates $grep_regex, which checks whether a given version is installed, to
match against the two different output formats of `pip list`. No longer
relies on pip freeze.

Additionally, this commit:

Removes the multiple exec resources which were mostly the same and replaces
with variables $pip_exec_name, $command, and $unless_command, which get
used in a single exec resource at the end of the manifest.

Checks if $pkgname contains ==, and if so splits based on == and sets
$real_pkgname to the first part, and $ensure variable to the second part.
Uses of $pkgname have been replaced with $real_pkgname.

#### This Pull Request (PR) fixes the following issues
Fixes #193

bastelfreak commented:
Hi @wolttam, thanks for the PR. Can you add an acceptance test to verify that the bug is fixed? Please also have a look at the failed travis jobs.

wolttam commented:
I fixed what was causing the existing tests to fail -- missing quotes (oops).

And, I added a new test for this, except, here's a quandary... git (or some other vcs, but git seems to be the appropriate choice) is required for it to run. I'm quite unfamiliar with the test framework in use here, so I'm not sure how to make git available. I'm looking into it, though.

bastelfreak commented:
this will install git within the docker environment:

```patch
diff --git a/spec/acceptance/virtualenv_spec.rb b/spec/acceptance/virtualenv_spec.rb
index a3520c3..dbcfd6d 100644
--- a/spec/acceptance/virtualenv_spec.rb
+++ b/spec/acceptance/virtualenv_spec.rb
@@ -113,21 +113,22 @@ describe 'python class' do
     end
     it 'works with editable=>true' do
       pp = <<-EOS
-      class { 'python' :
+      package{'git':
+        ensure => 'present',
+      }
+      -> class { 'python' :
         version    => 'system',
         pip        => 'present',
         virtualenv => 'present',
       }
-      ->
-      python::virtualenv { 'venv' :
+      -> python::virtualenv { 'venv' :
         ensure     => 'present',
         systempkgs => false,
         venv_dir   => '/opt/venv5',
         owner      => 'root',
         group      => 'root',
       }
-      ->
-      python::pip { 'rpyc' :
+      -> python::pip { 'rpyc' :
         ensure     => '4.1.0',
         url        => 'git+https://github.com/tomerfiliba/rpyc.git',
         editable   => true,

```

wolttam commented:
> this will install git within the docker environment (and improves the formatting):
> ....
Of course! Added that, pushing now.

bastelfreak commented on manifests/pip.pp:
```diff
@@ -135,14 +135,21 @@
+  $real_pkgname = $pkgname
+  if $pkgname =~ /==/ {
+    $parts = split($pkgname, '==')
+    $real_pkgname = $parts[0]
```
you reassign `$real_pkgname`. This should not work. The assignment in line 128 should probably happen in an else block. Can you also add a test where the pkgname contains a `==`?

wolttam replied:
Made this fix, and added the test.

bastelfreak commented on spec/acceptance/virtualenv_spec.rb:
```diff
@@ -111,5 +103,57 @@ class { 'python' :
+    end
+    it 'works with == in pkgname' do
+      pp = <<-EOS
+      'python' :
```
I think something got lost here?

```suggestion
      class{'python' :
```

wolttam replied:
Ouch! While I'm add it, would it be preferred if I removed the leading space before all resource names?

Also, I found what's causing the first two tests to fail: using $path in the exec instead of $_path, going to push that along with this correction.

bastelfreak replied:
Removing the leading space isn't required and I think it would just make the diff harder to read. I think we can keep that as is.

bastelfreak commented:
Thanks for the awesome work!
